src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -0,0 +1,355 @@
+        public static bool operator !=(Tensor<T> left, T right) => !(left == right);
+
+        public sealed class Enumerator : IEnumerator<T>
+        {
+            /// <summary>The span being enumerated.</summary>
+            private readonly Tensor<T> _tensor;
+            private nint[] _curIndices;
+            /// <summary>The total item count.</summary>
+            private nint _items;
+
+            /// <summary>Initialize the enumerator.</summary>
+            /// <param name="tensor">The tensor to enumerate.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal Enumerator(Tensor<T> tensor)
+            {
+                _tensor = tensor;
+                _items = -1;
+                _curIndices = new nint[_tensor.Rank];
+                _curIndices[_tensor.Rank - 1] = -1;
+            }
+
+            /// <summary>Advances the enumerator to the next element of the span.</summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public bool MoveNext()
+            {
+                SpanHelpers.AdjustIndices(_tensor.Rank - 1, 1, ref _curIndices, _tensor.Lengths);
+
+                _items++;
+                return _items < _tensor.LinearLength;
+            }
+
+            public void Reset()
+            {
+                Array.Clear(_curIndices);
+                _curIndices[_tensor.Rank - 1] = -1;
+            }
+
+            public void Dispose()
+            {
+
+            }
+
+            T IEnumerator<T>.Current => _tensor[_curIndices];
+
+            object IEnumerator.Current => _tensor[_curIndices];
+        }
+
+        // REVIEW: PROBABLY REMOVE.
+        public override bool Equals(object? obj)
+        {
+            return Equals(obj as Tensor<T>);
+        }
+
+        // REVIEW: PROBABLY REMOVE.
+        public override int GetHashCode()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Get a string representation of the tensor.
+        /// </summary>
+        private string ToMetadataString()
+        {
+            var sb = new StringBuilder("[");
+
+            int n = Rank;
+            if (n == 0)
+            {
+                sb.Append(']');
+            }
+            else
+            {
+                for (int i = 0; i < n; i++)
+                {
+                    sb.Append(Lengths[i]);
+                    if (i + 1 < n)
+                        sb.Append('x');
+                }
+
+                sb.Append(']');
+            }
+            sb.Append($", type = {typeof(T)}, isPinned = {IsPinned}");
+
+            return sb.ToString();
+        }
+
+        // REVIEW: This will be renamed.
+        public string ToCSharpString()
+        {
+            var sb = new StringBuilder();
+            sb.AppendLine(ToMetadataString());
+            sb.AppendLine("{");
+            sb.Append(AsSpan().ToCSharpString());
```

Comment:
Each of these ToMetadataString and ToCSharpString calls are allocating possibly large strings, just to turn around and append them to another StringBuilder. If they need to be factored into separate methods like this, it'd be better to have the helpers accept a StringBuilder to write into, and just write directly into the caller's builder.